gui/src/ruler.cc

```diff
@@ -0,0 +1,362 @@
+            segmentScreenSize = floor(mapRange(majorInterval, 0.0, upperLimit - lowerLimit, 0.0, rulerSize));
+
+            // If we've found a segment of appropriate size, we can stop
+            if (segmentScreenSize >= MIN_SEGMENT_SIZE) { break; }
+        }
+
+        // Try the next interval
+        intervalIndex++;
+        if (intervalIndex == VALID_INTERVALS.size())
+        {
+            // We tried all intervals for the current n
+            intervalIndex = 0;
+            intervalN++;
+        }
+    }
+}
+
+double Ruler::mapRange(double x, double a_lower, double a_upper, double b_lower, double b_upper)
+{
+    double a_width = a_upper - a_lower;
+    double b_width = b_upper - b_lower;
+    double scale = b_width / a_width;
+
+    return b_lower + scale * (x - a_lower);
+}
+
+gboolean Ruler::drawCallback(GtkWidget *widget, cairo_t *cr, gpointer data)
+{
+
+    auto *ruler = static_cast<Ruler *>(data);
+
+    if (ruler->drawingArea == nullptr) { return FALSE; }
+
+    double width = ruler->width;
+    double height = ruler->height;
+
+    // Draw background
+    GtkStyleContext *context = gtk_widget_get_style_context(widget);
+    gtk_render_background(context, cr, 0, 0, ruler->width, ruler->height);
+
+    // Draw outline along left and right sides and along the bottom
+    gdk_cairo_set_source_rgba(cr, &(ruler->lineColor));
+
+    cairo_set_line_width(cr, Ruler::LINE_WIDTH);
+    if (ruler->orientation == HORIZONTAL)
+    {
+        cairo_move_to(cr, LINE_COORD_OFFSET, 0);
+        cairo_line_to(cr, LINE_COORD_OFFSET, height);
+
+        cairo_move_to(cr, width - LINE_COORD_OFFSET, 0);
+        cairo_line_to(cr, width - LINE_COORD_OFFSET, height);
+        cairo_stroke(cr);
+
+        cairo_set_line_width(cr, 2 * Ruler::LINE_WIDTH);
+        cairo_move_to(cr, 0, height - LINE_COORD_OFFSET);
+        cairo_line_to(cr, width, height - LINE_COORD_OFFSET);
+        cairo_stroke(cr);
+    } else if (ruler->orientation == VERTICAL)
+    {
+        cairo_move_to(cr, 0, LINE_COORD_OFFSET);
+        cairo_line_to(cr, width, LINE_COORD_OFFSET);
+
+        cairo_move_to(cr, 0, height - LINE_COORD_OFFSET);
+        cairo_line_to(cr, width, height - LINE_COORD_OFFSET);
+        cairo_stroke(cr);
+
+        cairo_set_line_width(cr, 2 * Ruler::LINE_WIDTH);
+        cairo_move_to(cr, width - LINE_COORD_OFFSET, 0);
+        cairo_line_to(cr, width - LINE_COORD_OFFSET, height);
+        cairo_stroke(cr);
+    }
+    cairo_set_line_width(cr, Ruler::LINE_WIDTH);
+
+    // Calculate the line length for the major ticks given the size of the ruler
+    double lineLength = NAN;
+    if (ruler->orientation == HORIZONTAL)
+    {
+        lineLength = Ruler::MAJOR_TICK_LENGTH * height;
+    }
+    else
+    {
+        lineLength = Ruler::MAJOR_TICK_LENGTH * width;
+    }
+
+    // Draw positive side of the ruler
+    if (ruler->upperLimit > 0) // If part of the range is indeed positive
+    {
+        // Draw the range [max(0, lowerLimit), upperLimit]
```

Comment:
Why is it that we can't draw the entire range in one go?

Reply:
I would put that complexity in a function that computes the first tick mark. Something like
```
int firstTickMark(int low, int tickDistance)
{
   if(low < 0) return low/tickDistance;
   return (low+tickDistance-1)/tickDistance;
}
```
Then similarly for the last tick, and then start from there...